homeassistant/components/panasonic_viera/__init__.py

```diff
@@ -228,11 +291,84 @@ async def _handle_errors(self, func, *args):
             _LOGGER.error(
                 "The connection couldn't be encrypted. Please reconfigure your TV"
             )
-        except (TimeoutError, URLError, SOAPError, OSError):
-            self.state = STATE_OFF
+        except (TimeoutError, URLError, SOAPError, OSError) as err:
+            _LOGGER.debug("Could not establish remote connection: %s", err)
+            if self._control is not None:
+                await self.shutdown()
             self.available = self._on_action is not None
-            await self.async_create_remote_control()
+            self.connected = False
         except Exception as err:  # pylint: disable=broad-except
             _LOGGER.exception("An unknown error occurred: %s", err)
-            self.state = STATE_OFF
+            if self._control is not None:
+                await self.shutdown()
             self.available = self._on_action is not None
+            self.connected = False
+
+    async def on_event(self, service, properties):
+        """Parse and store received properties."""
+        if service is URL_EVENT_0_DMR:
+            if "Volume" in properties:
+                self._volume = properties["Volume"]["@val"]
```

Comment:
This looks like data parsing that should be part of the library.

Reply:
Yep, I agree. I'll add it to the library